File: Leetcode 3601-3700/3690/3690.py
# Leetcode 3690: Split and Merge Array Transformation
# https://leetcode.com/problems/split-and-merge-array-transformation/
# Solved on 29th of November, 2025
from collections import deque


class Solution:
    def minSplitMerge(self, nums1: list[int], nums2: list[int]) -> int:
        """
        Calculates the minimum number of split and merge operations to transform nums1 into nums2.

        Args:
            nums1 (list[int]): The starting array.
            nums2 (list[int]): The target array.
        Returns:
            int: The minimum number of operations, or -1 if transformation is not possible.
        """
        startState = tuple(nums1)
        targetState = tuple(nums2)

        if startState == targetState:
            return 0

        queue = deque([(startState, 0)])
        visited = {startState}
        n = len(nums1)

        while queue:
            currentState, currentDist = queue.popleft()

            for i in range(n):
                for j in range(i + 1, n + 1):
                    subArray = currentState[i:j]
                    remainingArray = currentState[:i] + currentState[j:]

                    for k in range(len(remainingArray) + 1):
                        nextState = remainingArray[:k] + subArray + remainingArray[k:]

                        if nextState == targetState:
                            return currentDist + 1

                        if nextState not in visited:
                            visited.add(nextState)
                            queue.append((nextState, currentDist + 1))

        return -1

File: Leetcode 3601-3700/3690/3690.java
// Leetcode 3690: Split and Merge Array Transformation
// https://leetcode.com/problems/split-and-merge-array-transformation/
// Solved on 29th of November, 2025
import java.util.*;

class Solution {
    /**
     * Calculates the minimum number of moves required to transform `nums1` into `nums2`
     * by repeatedly splitting a subarray and merging it back into a different position.
     * @param nums1 The initial array of integers.
     * @param nums2 The target array of integers.
     * @return The minimum number of moves, or -1 if `nums2` is unreachable.
     */
    public int minSplitMerge(int[] nums1, int[] nums2) {
        if (Arrays.equals(nums1, nums2)) {
            return 0;
        }

        int n = nums1.length;
        String target = Arrays.toString(nums2);
        
        Queue<int[]> queue = new ArrayDeque<>();
        Set<String> visited = new HashSet<>();
        
        queue.offer(nums1);
        visited.add(Arrays.toString(nums1));
        
        int moves = 0;
        
        while (!queue.isEmpty()) {
            moves++;
            int size = queue.size();
            
            for (int s = 0; s < size; s++) {
                int[] current = queue.poll();
                
                for (int i = 0; i < n; i++) {
                    for (int j = i; j < n; j++) {
                        int[] remainder = new int[n - (j - i + 1)];
                        int rIdx = 0;
                        for (int k = 0; k < n; k++) {
                            if (k < i || k > j) {
                                remainder[rIdx++] = current[k];
                            }
                        }
                        
                        for (int k = 0; k <= remainder.length; k++) {
                            if (k == i) continue;
                            
                            int[] next = new int[n];
                            int nIdx = 0;
                            
                            for (int p = 0; p < k; p++) {
                                next[nIdx++] = remainder[p];
                            }
                            
                            for (int p = i; p <= j; p++) {
                                next[nIdx++] = current[p];
                            }
                            
                            for (int p = k; p < remainder.length; p++) {
                                next[nIdx++] = remainder[p];
                            }
                            
                            String nextStr = Arrays.toString(next);
                            if (nextStr.equals(target)) {
                                return moves;
                            }
                            
                            if (visited.add(nextStr)) {
                                queue.offer(next);
                            }
                        }
                    }
                }
            }
        }
        
        return -1;
    }
}